### DISCUSSION TITLE
Enable CUDA on a Linux guest on Jetson TX2

### DISCUSSION THREAD

**Discussion by Alan-Kuan** [bao-project/bao-hypervisor#119] [General]
Hello.

I'm trying to enable CUDA on a Linux guest on Jetson TX2.
Is there anyone has experience in doing this?

I think there're 2 challenges:
1. Describe the GPU in the device tree and specify it in the VM config
2. Install relevant firmware on the Linux guest

**Comment by josecm** [comment]
Hey Alan!

I do not have much experience in using the GPU on the Jetson within a Bao system. I agree with your point 1. But I don't fully understand how point 2 impacts a bao configuration. Nevertheless, I believe a GPU can be seen purely as a passthrough device in Bao. So essentially defining the correct MMIO regions and interrupts for the target VM in the bao config file. An extra step that must be taken is to configure the SMMU accordingly (`.id` field in the `vm_dev_region`). For this you must first define the ID in the firmware, so you need to use our fork of atf and add it to the streamid table (see [ this commit](https://github.com/bao-project/arm-trusted-firmware/commit/7d31f572ba24e90a265c6cc370bb7270ea9841c4)). 

Let me know how this evolves

**Reply by Alan-Kuan** [reply]
Thanks for your advice. I'm not proficient in this area, but I'll try my best.

**Reply by josecm** [reply]
Feel free to ask for any more specific issue you may encounter.
